pkg/v2/tkr/controller/source/compatibility/reconciler.go

```diff
@@ -0,0 +1,189 @@
+
+type Config struct {
+	TKRNamespace string
+}
+
+func (r *Reconciler) SetupWithManager(mgr ctrl.Manager) error {
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&runv1.TanzuKubernetesRelease{}).
+		Watches(
+			&source.Kind{Type: &corev1.ConfigMap{}},
+			handler.EnqueueRequestsFromMapFunc(r.enqueueAllTKRs),
+			builder.WithPredicates(predicate.NewPredicateFuncs(func(o client.Object) bool {
+				return o.GetNamespace() == r.Config.TKRNamespace && o.GetName() == constants.BOMMetadataConfigMapName
+			}))).
+		Watches(
+			&source.Kind{Type: &clusterv1.Cluster{}},
+			handler.EnqueueRequestsFromMapFunc(r.enqueueAllTKRs),
+			builder.WithPredicates(hasManagementClusterRoleLabel, predicate.AnnotationChangedPredicate{})).
+		Named("tkr_compatibility").
+		Complete(r)
+}
+
+// enqueueAllTKRs returns reconcile.Request for all existing TKRs.
+func (r *Reconciler) enqueueAllTKRs(o client.Object) []ctrl.Request {
+	r.Log.Info("enqueueing all TKRs, triggered by object",
+		"kind", o.GetObjectKind().GroupVersionKind().Kind, "namespace", o.GetNamespace(), "name", o.GetName())
+
+	tkrs := &runv1.TanzuKubernetesReleaseList{}
+	if err := r.Client.List(r.Ctx, tkrs); err != nil {
+		r.Log.Error(err, "error listing TKRs")
+		return nil
+	}
+	result := make([]ctrl.Request, len(tkrs.Items))
+	for i := range tkrs.Items {
+		result[i].NamespacedName.Name = tkrs.Items[i].Name
+	}
+	return result
+}
+
+var hasManagementClusterRoleLabel = func() predicate.Predicate {
+	selector, _ := labels.Parse(constants.ManagementClusterRoleLabel)
+	return predicate.NewPredicateFuncs(func(o client.Object) bool {
+		return selector.Matches(labels.Set(o.GetLabels()))
+	})
+}()
+
+func (r *Reconciler) Reconcile(ctx context.Context, req ctrl.Request) (result ctrl.Result, retErr error) {
+	tkr := &runv1.TanzuKubernetesRelease{}
+	if err := r.Client.Get(ctx, req.NamespacedName, tkr); err != nil {
+		if apierrors.IsNotFound(err) {
+			return ctrl.Result{}, nil // do nothing if the ConfigMap does not exist
+		}
+		return ctrl.Result{}, err
+	}
+
+	ps := patchset.New(r.Client)
+	defer func() {
+		// apply patches unless an error is being returned
+		if retErr != nil {
+			return
+		}
+		if err := ps.Apply(ctx); err != nil {
+			if err = kerrors.FilterOut(err, apierrors.IsConflict); err == nil {
+				// retry if someone updated an object we wanted to patch
+				result = ctrl.Result{Requeue: true}
+			}
+			retErr = errors.Wrap(err, "applying patches to TKRs")
+		}
+	}()
+
+	ps.Add(tkr)
+
+	return ctrl.Result{}, r.updateTKRCompatibleCondition(ctx, tkr)
+}
+
+func (r *Reconciler) updateTKRCompatibleCondition(ctx context.Context, tkr *runv1.TanzuKubernetesRelease) error {
+	compatibleSet, err := r.getCompatibleSet(ctx)
+	if err != nil {
+		return err
+	}
+
+	if _, isCompatible := compatibleSet[tkr.Spec.Version]; isCompatible {
+		conditions.MarkTrue(tkr, runv1.ConditionCompatible)
+		return nil
+	}
+	conditions.MarkFalse(tkr, runv1.ConditionCompatible, "", clusterv1.ConditionSeverityInfo, "")
```

Comment:
As per the spec, the severity for `false` status for compatibility condition should be `Warning`

Reply:
Absolutely right! Fixing.

Reply:
Done.